src/dynamics/poincare_recurrence.lean

```diff
@@ -0,0 +1,169 @@
+
+lemma map_volume_eq (hf : volume_preserving f) : measure_space.μ.map f = measure_space.μ :=
+measure.ext $ λ s hs, by simpa only [measure.map_apply hf.measurable hs] using hf.2 hs
+
+lemma volume_eq (hf : volume_preserving f) {s : set α} (hs : is_measurable s) :
+  volume (f ⁻¹' s) = volume s :=
+hf.2 hs
+
+lemma comp (hf : volume_preserving f) {g : α → α} (hg : volume_preserving g) :
+  volume_preserving (f ∘ g) :=
+⟨hf.measurable.comp hg.measurable, λ s hs,
+  by rw [preimage_comp, hg.volume_eq (hf.measurable s hs), hf.volume_eq hs]⟩
+
+lemma iterate (hf : volume_preserving f) : ∀ n, volume_preserving (f^[n])
+| 0 := volume_preserving.id α
+| (n+1) := (iterate n).comp hf
+
+lemma exists_mem_image_mem_of_volume_lt_mul_volume (hf : volume_preserving f) {s : set α}
+  (hs : is_measurable s) {n : ℕ} (hvol : volume (univ : set α) < n * volume s) :
+  ∃ (x ∈ s) (m ∈ Ioo 0 n), f^[m] x ∈ s :=
+begin
+  have A : ∀ m, is_measurable (f^[m] ⁻¹' s) := λ m, (hf.iterate m).measurable s hs,
+  have B : ∀ m, volume (f^[m] ⁻¹' s) = volume s, from λ m, (hf.iterate m).volume_eq hs,
+  have : volume (univ : set α) < (finset.range n).sum (λ m, volume (f^[m] ⁻¹' s)),
+    by simpa [B, add_monoid.smul_eq_mul],
+  rcases exists_nonempty_inter_of_volume_univ_lt_sum_volume (λ m hm, A m) this
+    with ⟨i, hi, j, hj, hij, x, hxi, hxj⟩,
+  wlog hlt : i < j := lt_or_gt_of_ne hij using [i j, j i],
+  simp only [set.mem_preimage, finset.mem_range] at hi hj hxi hxj,
+  refine ⟨f^[i] x, hxi, j - i, ⟨nat.sub_pos_of_lt hlt, lt_of_le_of_lt (j.sub_le i) hj⟩, _⟩,
+  rwa [← nat.iterate_add, nat.sub_add_cancel (le_of_lt hlt)]
+end
+
+lemma exists_mem_image_mem (hf : volume_preserving f) {s : set α} (hs : is_measurable s)
+  (hs' : volume s ≠ 0) (hvol : volume (univ : set α) ≠ ⊤) :
+  ∃ (x ∈ s) (m > 0), f^[m] x ∈ s :=
+begin
+  rcases ennreal.exists_nat_mul_gt hs' hvol with ⟨N, hN⟩,
+  rcases hf.exists_mem_image_mem_of_volume_lt_mul_volume hs hN with ⟨x, hx, m, hm, hmx⟩,
+  exact ⟨x, hx, m, hm.1, hmx⟩
+end
+
+lemma is_measurable_not_recurrent (hf : measurable f) {s : set α} (hs : is_measurable s) (n : ℕ) :
+  is_measurable (s ∩ ⋂ m ≥ n, f^[m] ⁻¹' -s) :=
+hs.inter (is_measurable.Inter $ λ m, is_measurable.Inter_Prop $
+      λ hm, (hf.iterate m).preimage hs.compl)
+
+lemma volume_mem_forall_image_not_mem_eq_zero (hf : volume_preserving f) {s : set α}
+  (hs : is_measurable s) (n : ℕ) (hvol :  volume (univ : set α) ≠ ⊤) :
+  volume (s ∩ ⋂ m ≥ n, f^[m] ⁻¹' -s) = 0 :=
+begin
+  by_contradiction H,
+  rcases exists_mem_image_mem (hf.iterate n)
+    (is_measurable_not_recurrent hf.measurable hs n) H hvol
+    with ⟨x, ⟨hxs, hxs'⟩, m, m0, ⟨hmx, hmx'⟩⟩,
+  simp only [← nat.iterate_mul, set.mem_inter_iff, set.mem_Inter, set.mem_preimage] at hxs' hmx,
+  exact hxs' (n * m) (nat.le_mul_of_pos_right m0) hmx
+end
+
+/-- Poincaré recurrence theorem: given a volume preserving map `f` and a measurable set `s`,
+almost every point `x ∈ s`-/
```

Comment:
This docstring seems to be incomplete

Reply:
@jcommelin Sorry for publishing this too early. The changes happened to be on the "poincare-recurrence" branch for historical reasons. Right now I'm looking for feedback only on my changes to `borel_space.lean`.